Sources/PackageLoading/ManifestLoader.swift

```diff
@@ -1115,3 +1066,150 @@ extension TSCBasic.Diagnostic.Message {
             """)
     }
 }
+
+/// SQLite backed persistent cache.
+private final class SQLiteManifestCache: Closable {
+    let fileSystem: FileSystem
+    let location: SQLite.Location
+
+    private var state = State.idle
+    private let stateLock = Lock()
+
+    private let diagnosticsEngine: DiagnosticsEngine?
+    private let jsonEncoder: JSONEncoder
+    private let jsonDecoder: JSONDecoder
+
+    init(location: SQLite.Location, diagnosticsEngine: DiagnosticsEngine? = nil) {
+        self.location = location
+        switch self.location {
+        case .path, .temporary:
+            self.fileSystem = localFileSystem
+        case .memory:
+            self.fileSystem = InMemoryFileSystem()
+        }
+        self.diagnosticsEngine = diagnosticsEngine
+        self.jsonEncoder = JSONEncoder.makeWithDefaults()
+        self.jsonDecoder = JSONDecoder.makeWithDefaults()
+    }
+
+    convenience init(path: AbsolutePath, diagnosticsEngine: DiagnosticsEngine? = nil) {
+        self.init(location: .path(path), diagnosticsEngine: diagnosticsEngine)
+    }
+
+    deinit {
+        guard case .disconnected = (self.stateLock.withLock { self.state }) else {
+            return assertionFailure("db should be closed")
```

Comment:
This will get compiled out for release builds (I think), but would it make sense to also close the database in `deinit` in this case as a safeguard in release builds?  That could help mitigate the symptoms even if there's a bug here.